state_synchronizer/src/coordinator.rs

```diff
@@ -336,4 +350,30 @@ impl<T: ExecutorProxyTrait> SyncCoordinator<T> {
         req.set_ledger_info_with_sigs(ledger_info.into_proto());
         self.executor_proxy.execute_chunk(req).await
     }
+
+    async fn check_subscriptions(&mut self) -> Result<()> {
+        let ledger_info = self.executor_proxy.get_latest_ledger_info().await?;
+        let committed_version = self.known_version;
+        let mut ready = HashMap::new();
+
+        self.subscriptions
+            .retain(|peer_id, (expiry, known_version, limit)| {
+                // filter out expired peer requests
+                if SystemTime::now().duration_since(expiry.clone()).is_ok() {
+                    return false;
+                }
+                if *known_version < committed_version {
+                    ready.insert(peer_id.clone(), (*known_version, *limit));
+                }
+                // retain non-ready requests
+                *known_version >= committed_version
+            });
+
+        for (peer_id, (known_version, limit)) in ready.into_iter() {
```

Comment:
this has a head-of-line blocking problem by enforcing things to finish in order. Should we spawn all the futures first and then wait for all of them to finish using a `FuturesUnordered`?

Reply:
So in theory all of them are doing same type of work. I'd expect that even response will be the same for all subscribers, so it's not a big issue
But it's better for latency to process it in parallel anyway. Thanks for flagging it!